linfa-linear/src/elastic.rs

```diff
@@ -0,0 +1,572 @@
+    }
+
+    /// Set the overall parameter penalty parameter of the elastic net.
+    /// Use `l1_ratio` to configure how the penalty distributed to L1 and L2
+    /// regularization.
+    pub fn penalty(mut self, penalty: F) -> Self {
+        self.penalty = penalty;
+        self
+    }
+
+    /// Set l1_ratio parameter of the elastic net. Controls how the parameter
+    /// penalty is distributed to L1 and L2 regularization.
+    /// Setting `l1_ratio` to 1.0 is equivalent to a "Lasso" penalization,
+    /// setting it to 0.0 is equivalent to "Ridge" penalization.
+    ///
+    /// Defaults to `0.5` if not set
+    ///
+    /// `l1_ratio` must be between `0.0` and `1.0`.
+    pub fn l1_ratio(mut self, l1_ratio: F) -> Self {
+        if l1_ratio < F::zero() || l1_ratio > F::one() {
+            panic!("Invalid value for l1_ratio, needs to be between 0.0 and 1.0");
+        }
+        self.l1_ratio = l1_ratio;
+        self
+    }
+
+    /// Configure the elastic net model to fit an intercept.
+    /// Defaults to `true` if not set.
+    pub fn with_intercept(mut self, with_intercept: bool) -> Self {
+        self.with_intercept = with_intercept;
+        self
+    }
+
+    /// Set the tolerance which is the minimum absolute change in any of the
+    /// model parameters needed for the parameter optimization to continue.
+    ///
+    /// Defaults to `1e-4` if not set
+    pub fn tolerance(mut self, tolerance: F) -> Self {
+        self.tolerance = tolerance;
+        self
+    }
+
+    /// Set the maximum number of iterations for the optimization routine.
+    ///
+    /// Defaults to `1000` if not set
+    pub fn max_iterations(mut self, max_iterations: u32) -> Self {
+        self.max_iterations = max_iterations;
+        self
+    }
+
+    /// Fit an elastic net model given a feature matrix `x` and a target
+    /// variable `y`.
+    ///
+    /// The feature matrix `x` must have shape `(n_samples, n_features)`
+    ///
+    /// The target variable `y` must have shape `(n_samples)`
+    ///
+    /// Returns a `FittedElasticNet` object which contains the fitted
+    /// parameters and can be used to `predict` values of the target variable
+    /// for new feature values.
+    pub fn fit(&self, x: &Array2<F>, y: &Array1<F>) -> Result<FittedElasticNet<F>, String> {
+        let (intercept, y) = self.compute_intercept(y);
+        let (parameters, duality_gap, n_steps) = coordinate_descent(
+            &x,
+            &y,
+            self.tolerance,
+            self.max_iterations,
+            self.l1_ratio,
+            self.penalty,
+        );
+        Ok(FittedElasticNet {
+            intercept,
+            parameters,
+            duality_gap,
+            n_steps,
+        })
+    }
+
+    /// Compute the intercept as the mean of `y` and center `y` if an intercept should
+    /// be used, use `0.0` as intercept and leave `y` unchanged otherwise.
+    fn compute_intercept<'a>(&self, y: &'a Array1<F>) -> (F, Cow<'a, Array1<F>>) {
```

Comment:
I think there is a `ndarray::CowArrray` for this usecase

Reply:
Thanks for pointing this out. I've tried changing the code to using CowArray, but in this case I found it inconvenient. 

CowArray wraps either a `ArrayBase<ViewRepr, ...>` or an `ArrayBase<OwnedRepr, ...>`, which is semantically the same as Cow, but slightly different: For the view case, I have to take a slice and then convert that into the owned array. Then the subsequent code takes a shared ref to an owned array (this could of course be changed as well....). In the end I found this inconvenient (in this instance) and left this like it is.

Reply:
wait when the coordinate descent routine takes the intercept as reference to an owned one, the `Cow` wrapper is always converted to an owned object. Why can't you then not just return ` -> (F, Array1<F>)` and clone the intercept for the second case implicitly?